### DISCUSSION TITLE
Support `devEngines` in package.json

### DISCUSSION THREAD

**Discussion by Jiralite** [vercel/turborepo#9473] [Ideas]
### Goals

1. Support `devEngines` in package.json

### Background

The [`devEngines`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#devengines) field in the package.json should be supported. Currently, only `packageManager` is scanned:

https://github.com/vercel/turborepo/blob/13bfef4024b12fd7163c2d46b06fa962b5168bbd/crates/turborepo-repository/src/package_manager/mod.rs#L455-L458

I noticed pnpm is used here. pnpm also does not support `devEngines` as of now: https://github.com/pnpm/pnpm/issues/8153

### Proposal

During this period, I believe both should be checked with a priority of `devEngines`.

**Comment by anthonyshew** [comment]
We are definitely excited to upgrade from `packageManager` to `devEngines` along with the rest of the ecosystem.

Has `devEngines` been officially accepted/implemented into Node.js? I see the pnpm issue you've linked. What about npm and yarn?

We will need to come up with a sensible migration strategy to handle the matrix of `packageManager` || `devEngines` && package manager && `turbo` but I don't believe we will have a problem doing so.

**Reply by Jiralite** [reply]
npm supports this (https://github.com/npm/cli/pull/7766). For the others, I found these which are great to read for context:

- https://github.com/nodejs/corepack/issues/567
- https://github.com/nodejs/node/pull/51981 (Definitely check this out. A Node.js member opened this)
- https://github.com/nodejs/package-maintenance/issues/609
- https://github.com/nodejs/package-maintenance/pull/594
- https://github.com/yarnpkg/berry/issues/6443

So ultimately, it seems adoption is still in progress with Node.js seemingly planning to remove corepack in favour of `devEngines`.

**Reply by anthonyshew** [reply]
Great, thank you for finding those. We'll be looking to this when we can.

**Comment by turboRails** [comment]
Continued here https://github.com/vercel/turborepo/discussions/10848

**Comment by sirkostya009** [comment]
in recent versions (10.14 if i remember correctly) pnpm added support for `devEngines.runtime`, and currently if you try to `turbo prune --docker` and have (for example) node runtime, the generated `out/json/pnpm-lock.yaml` would not include `node@runtime` package key, causing `pnpm i --frozen-lockfile` installs to fail!
